docker/serverless/gcloud_build_image

```diff
@@ -12,31 +12,65 @@
 #   gcloud endpoints services list
 #   gcloud endpoints configs list --service=${SERVICE}
 # Use the latest one for the CONFIG_ID
+#
+# The script will use the latest ESPv2 version by default.
+# Use the -v option to pass in a custom version (example: -v 2.9).
 
-# Default to the release image if not specified
-BASE_IMAGE=gcr.io/endpoints-release/endpoints-runtime-serverless:2
+# Default to the latest released ESPv2 version.
+BASE_IMAGE_NAME="gcr.io/endpoints-release/endpoints-runtime-serverless"
+ESP_TAG="2"
 
 function error_exit() {
   # ${BASH_SOURCE[1]} is the file name of the caller.
   echo "${BASH_SOURCE[1]}: line ${BASH_LINENO[0]}: ${1:-Unknown Error.} (exit ${2:-1})" 1>&2
   exit ${2:-1}
 }
 
-while getopts :c:s:p:i: arg; do
+while getopts :c:s:p:v:i: arg; do
   case ${arg} in
     c) CONFIG_ID="${OPTARG}";;
     s) SERVICE="${OPTARG}";;
     p) PROJECT="${OPTARG}";;
-    i) BASE_IMAGE="${OPTARG}";;
+    v) ESP_TAG="${OPTARG}";;
+    i)
+      BASE_IMAGE="${OPTARG}"
+      ESP_FULL_VERSION="custom"
+      ;;
     \?) error_exit "Unrecognized argument -${OPTARG}";;
   esac
 done
 
 [[ -n "${PROJECT}" ]] || error_exit "Missing required PROJECT"
 [[ -n "${SERVICE}" ]] || error_exit "Missing required SERVICE"
 [[ -n "${CONFIG_ID}" ]] || error_exit "Missing required CONFIG_ID"
+
+# If user did not pass in custom image, then form the fully-qualified base image.
+if [ -z "${BASE_IMAGE}" ]; then
+  BASE_IMAGE="${BASE_IMAGE_NAME}:${ESP_TAG}"
+fi;
 echo "Using base image: ${BASE_IMAGE}"
 
+# If user did not pass in a custom image, then determine the ESP version.
+if [ -z "${ESP_FULL_VERSION}" ]; then
+  echo "Determining fully-qualified ESP version for tag: ${ESP_TAG}"
+
+  ALL_TAGS=$(gcloud container images list-tags "${BASE_IMAGE_NAME}" \
+      --filter="tags~^${ESP_TAG}$" \
+      --format="value(tags)")
+  IFS=',' read -ra TAGS_ARRAY <<< "${ALL_TAGS}"
+
+  if [ ${#TAGS_ARRAY[@]} -eq 0 ]; then
+    error_exit "Did not find ESP version: ${ESP_TAG}"
+  fi;
+
+  # Get the longest tag by creating an array indexed by tag length. Then use the last element.
```

Comment:
Sort array by dictionary order? Otherwise, 2.10.0 and 2.10.1 are likely be out of order according to the comment though we haven't published minor version so far.

Reply:
This would never happen. If we publish 2.10.1, then tag 2 and 2.10 will only point to 2.10.1. It would never get a string with both 2.10.0 and 2.10.1 at the same time.

Reply:
OK, I see you query the tags of the image pointed by tag "2", which only gives 2, 2.X and 2.X.Y uniquely.